lti_consumer/signals.py

```diff
@@ -0,0 +1,27 @@
+"""
+LTI Consumer related Signal handlers
+"""
+
+from django.db.models.signals import post_save
+from django.dispatch import receiver
+
+from lti_consumer.models import LtiAgsScore
+from lti_consumer.plugin import compat
+
+
+@receiver(post_save, sender=LtiAgsScore, dispatch_uid='publish_grade_on_score_update')
+def publish_grade_on_score_update(sender, instance, **kwargs):  # pylint: disable=unused-argument
+    """
+    Publish grade to xblock whenever score saved/updated and its grading_progress is set to FullyGraded.
+    """
+    if instance.grading_progress == LtiAgsScore.FULLY_GRADED:
+        block = compat.load_block_as_anonymous_user(instance.line_item.resource_link_id)
```

Comment:
This is not working for me, I'm getting a 500 error when doing the request:
```
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/db/models/query.py", line 422, in create
    obj.save(force_insert=True, using=self.db)
  File "/edx/src/xblock-lti-consumer/lti_consumer/models.py", line 308, in save
    super().save(*args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/db/models/base.py", line 743, in save
    self.save_base(using=using, force_insert=force_insert,
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/db/models/base.py", line 791, in save_base
    post_save.send(
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/dispatch/dispatcher.py", line 173, in send
    return [
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/dispatch/dispatcher.py", line 174, in <listcomp>
    (receiver, receiver(signal=self, sender=sender, **named))
  File "/edx/src/xblock-lti-consumer/lti_consumer/signals.py", line 18, in publish_grade_on_score_update
    block = compat.load_block_as_anonymous_user(instance.line_item.resource_link_id)
  File "/edx/src/xblock-lti-consumer/lti_consumer/plugin/compat.py", line 43, in load_block_as_anonymous_user
    get_module_for_descriptor_internal(
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/courseware/module_render.py", line 874, in get_module_for_descriptor_internal
    (system, student_data) = get_module_system_for_user(
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/courseware/module_render.py", line 768, in get_module_system_for_user
    field_data = DateLookupFieldData(descriptor._field_data, course_id, user)  # pylint: disable=protected-access
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/edx_when/field_data.py", line 51, in __init__
    self._load_dates(course_id, user, use_cached=use_cached)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/edx_when/field_data.py", line 59, in _load_dates
    for (location, field), date in api.get_dates_for_course(course_id, user, use_cached=use_cached).items():
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/edx_when/api.py", line 142, in get_dates_for_course
    allow_relative_dates = _are_relative_dates_enabled(course_id)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/edx_when/api.py", line 57, in _are_relative_dates_enabled
    return RELATIVE_DATES_FLAG.is_enabled(course_key)
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/experiments/flags.py", line 253, in is_enabled
    return self.get_bucket(course_key) != 0
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/experiments/flags.py", line 165, in get_bucket
    if not hasattr(request, 'user') or not request.user.id:
AttributeError: 'NoneType' object has no attribute 'id'
[14/Nov/2020 21:50:29] "POST /api/lti_consumer/v1/lti/1/lti-ags/1/scores HTTP/1.1" 500 126762
```

Loading the block here (on signals.py) is erroring out for me. Can you check what's causing this?

Reply:
@giovannicimolin, It seems to be causing from ``django-crum`` module. As per documentation [here](https://github.com/ninemoreminutes/django-crum/blob/eeb350fc963360633b70472f06ddc6201ece5387/docs/index.rst#get_current_user) it returns ``user=None`` when not in request scope. I couldn't figure out why xblock signal is outside the request scope, when some other places seems to be working fine like [here](https://github.com/edx/edx-platform/blob/9cf2f9f298e5e8be3b3abcaadaf0b7a96d0de0df/common/djangoapps/course_modes/signals.py#L65).

Anyhow, I fixed the issue by [impersonating](https://github.com/ninemoreminutes/django-crum/blob/eeb350fc963360633b70472f06ddc6201ece5387/docs/index.rst#impersonateusernone) as ``AnonymousUser`` [here](https://github.com/open-craft/xblock-lti-consumer/blob/09c1e0a310ac5a1971dc1177334086762bee5971/lti_consumer/plugin/compat.py#L46). But I'm not sure if this is the right way to do it. Adding this to Author's concern section.

Reply:
This solution works and given it is present in the official documentation, I don't see a reason not to use it.